renovation_core_graphql/tags/resolvers.py

```diff
@@ -3,31 +3,69 @@
 import frappe
 from frappe.desk.doctype.tag.tag import add_tag, remove_tag
 from frappe.utils import cint
+from frappe_graphql.api import get_query
 
 
 def add_tag_link_resolver(obj, info: GraphQLResolveInfo, **kwargs):
     """
     Add a tag to a specific document.
     Adding the same tag again is a no-op.
     """
-    input = kwargs.get("input")
-    doctype = input.get("doctype")
-    docname = input.get("docname")
-    tag = input.get("tag")
-    return add_tag(tag=tag, dt=doctype, dn=docname)
+    r = frappe._dict(success_count=0, failure_count=0, total_count=len(kwargs.get("input")))
+    error_outputs = []
+    for tag_input in kwargs.get("input"):
+        try:
+            doctype = tag_input.get("doctype")
+            docname = tag_input.get("docname")
+            tag = tag_input.get("tag")
+            add_tag(tag=tag, dt=doctype, dn=docname)
+            frappe.db.commit()
+            r.success_count += 1
+        except Exception as e:
+            frappe.db.rollback()
+            r.failure_count += 1
+            error_outputs.append(e)
+    if len(error_outputs):
+        query, variables, operation_name = get_query()
+        log_tag_link_errors(query, variables, operation_name,
+                            {
+                                "success_count": r.success_count,
+                                "failure_count": r.failure_count,
+                                "total_count": r.total_count
+                            },
+                            error_outputs)
+    return r
 
 
 def remove_tag_link_resolver(obj, info: GraphQLResolveInfo, **kwargs):
     """
     Remove a tag from a specific document.
     Removing a tag that does not exist is a no-op.
     """
-    input = kwargs.get("input")
-    doctype = input.get("doctype")
-    docname = input.get("docname")
-    tag = input.get("tag")
-    remove_tag(tag=tag, dt=doctype, dn=docname)
-    return True
+    r = frappe._dict(success_count=0, failure_count=0, total_count=len(kwargs.get("input")))
```

Comment:
@e-lobo I know that adding a tag that already exists is a no-op. But would it improve the performance if the tag was checked against a document before adding it?

Reply:
@assemmarwan its done [internally](https://github.com/frappe/frappe/blob/86f3e45956152cf389500a6e4d7f7fb917ca1aef/frappe/desk/doctype/tag/tag.py#L77)